File: 04-03/tads/priorityqueue/MinHeap.java
package tads.priorityqueue;

public class MinHeap<E, P extends Comparable<P>> implements PriorityQueue<E, P> {
  private Object[] heap;
  private int heapSize;

  class Pair<E, P> {
    E fst;
    P snd;

    Pair(E fst, P snd) {
      this.fst = fst;
      this.snd = snd;
    }
  }

  @Override
  public void push(E element, P priority) {
    heapSize++;
    if (heap.length <= heapSize) {
      Object[] newHeap = new Object[heap.length * 2];
      for (int i = 1; i < heap.length; i++) {
        newHeap[i] = heap[i];
      }
      heap = newHeap;
    }

    Pair<E, P> pair = new Pair<E, P>(element, priority);
    heap[heapSize] = pair;

    siftUp(heapSize);
  }

  public void siftUp(int pos) {
    int parentPos = pos / 2;
    if (parentPos < 1) {
      return;
    }

    Pair<E, P> parent = (Pair<E, P>) heap[parentPos];
    Pair<E, P> node = (Pair<E, P>) heap[pos];

    if (parentIsGreater(node.snd, parent.snd)) {
      swap(pos, parentPos);
      siftUp(parentPos);
    }
  }

  private void swap(int a, int b) {
    Object aux = heap[a];
    heap[a] = heap[b];
    heap[b] = aux;
  }

  private boolean parentIsGreater(P child, P parent) {
    return parent.compareTo(child) > 0;
  }

  @Override
  public E pop() {
    // TODO Auto-generated method stub
    throw new UnsupportedOperationException("Unimplemented method 'pop'");
  }

  @Override
  public E top() {
    // TODO Auto-generated method stub
    throw new UnsupportedOperationException("Unimplemented method 'top'");
  }

  @Override
  public boolean isEmpty() {
    // TODO Auto-generated method stub
    throw new UnsupportedOperationException("Unimplemented method 'isEmpty'");
  }

  @Override
  public int size() {
    // TODO Auto-generated method stub
    throw new UnsupportedOperationException("Unimplemented method 'size'");
  }

}

File: 04-03/tads/priorityqueue/minHeap.cpp
#pragma once
#include "priorityQueue.cpp";

template <class E, class P>
struct MinHeapPair
{
  E fst;
  P snd;

  MinHeapPair(E fst, P snd)
  {
    this.fst = fst;
    this.snd = snd;
  }
};

template <class E, class P>
class MinHeap : PriorityQueue<E, P>
{
private:
  MinHeapPair<E, P> **heap;
  // cantidad de elementos en el heap
  int heapSize;
  // tamanio del array
  int heapLength;

  void swap(int a, int b)
  {
    MinHeapPair<E, P> *aux = heap[a];
    heap[a] = heap[b];
    heap[b] = aux;
  }

  bool parentIsGreater(P child, P parent)
  {
    return parent > child;
  }

public:
  void push(E element, P priority) override
  {
    heapSize++;
    if (heapLength <= heapSize)
    {
      MinHeapPair **newHeap = new MinHeapPair *[heapLength * 2];
      for (int i = 1; i < heapLength; i++)
      {
        newHeap[i] = heap[i];
      }
      heap = newHeap;
      heapLength = heapLength * 2;
    }

    MinHeapPair<E, P> *pair = new MinHeapPair<E, P>(element, priority);
    heap[heapSize] = pair;

    siftUp(heapSize);
  }

  void siftUp(int pos)
  {
    int parentPos = pos / 2;
    if (parentPos < 1)
    {
      return;
    }

    MinHeapPair<E, P> *parent = heap[parentPos];
    MinHeapPair<E, P> *node = heap[pos];

    if (parentIsGreater(node->snd, parent->snd))
    {
      swap(pos, parentPos);
      siftUp(parentPos);
    }
  }

  E pop() override
  {
    // TODO Auto-generated method stub
  }

  E top() override
  {
    // TODO Auto-generated method stub
  }

  bool isEmpty() override
  {
    // TODO Auto-generated method stub
  }

  int size() override
  {
    // TODO Auto-generated method stub
  }
};
